File: ruby/run-length-encoding/run_length_encoding.rb
require "english"

module RunLengthEncoding
  CONSECUTIVE_DATA_ELEMENTS = /([A-Za-z\s])\1+/.freeze
  private_constant :CONSECUTIVE_DATA_ELEMENTS
  RUN_LENGTH_ENCODING = /(?<count>\d+)(?<character>\D)/.freeze
  private_constant :RUN_LENGTH_ENCODING

  module_function

  def encode(input)
    return "" if input.empty?

    input.gsub(CONSECUTIVE_DATA_ELEMENTS) do
      "#{$MATCH.length}#{$LAST_PAREN_MATCH}"
    end
  end

  def decode(input)
    return "" if input.empty?

    input.gsub(RUN_LENGTH_ENCODING) do
      $LAST_MATCH_INFO[:character] * $LAST_MATCH_INFO[:count].to_i
    end
  end
end

File: javascript/run-length-encoding/run-length-encoding.js
// NOTE: \1 backreferences the single character match in parentheses.
// The parentheses only serve to provide a match for the backreference.
const CONSECUTIVE_DATA_ELEMENTS = /([\w\s])\1+/g
const RUN_LENGTH_ENCODING = /(\d+)(\D)/g

export const encode = input => {
  if (input === "") {
    return input
  }

  return input.replace(CONSECUTIVE_DATA_ELEMENTS, compress)
}

export const decode = input => {
  if (input === "") {
    return input
  }

  return input.replace(RUN_LENGTH_ENCODING, reconstruct)
}

function compress(match, character) {
  return `${match.length}${character}`
}

function reconstruct(_match, digit, character) {
  return character.repeat(digit)
}
